src/coalescenceml/artifact_store/base_artifact_store.py

```diff
@@ -0,0 +1,196 @@
+import textwrap
+from abc import ABC
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+)
+
+from pydantic import root_validator
+from tfx.dsl.io.fileio import NotFoundError
+
+from coalescenceml.artifact_store.exceptions import ArtifactStoreInterfaceError
+from coalescenceml.enums import StackComponentFlavor
+from coalescenceml.stack import StackComponent
+
+
+PathType = Union[bytes, str]
+
+
+# Bullshit hack b/c of TFX....
+def _catch_not_found_error(_func: Callable[..., Any]) -> Callable[..., Any]:
+    """Utility decorator used for catching a `FileNotFoundError` and
+    converting it to a `NotFoundError` in order to deal with the TFX exception
+    handling."""
+
+    def inner_function(*args: Any, **kwargs: Any) -> Any:
+        try:
+            return _func(*args, **kwargs)
+        except FileNotFoundError as e:
+            raise NotFoundError() from e
+
+    return inner_function
+
+
+class BaseArtifactStore(StackComponent, ABC):
+    """Base class for all CoalescenceML artifact stores.
+    Attributes:
+        path: The root path of the artifact store.
+    """
+
+    path: str
+
+    # Class Configuration
+    TYPE: ClassVar[StackComponentFlavor] = StackComponentFlavor.ARTIFACT_STORE
+    FLAVOR: ClassVar[str]
+    SUPPORTED_SCHEMES: ClassVar[Set[str]]
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initiate the Pydantic object and register the corresponding
+        filesystem."""
+        super(BaseArtifactStore, self).__init__(*args, **kwargs)
+        self._register()
+
+    @staticmethod
+    def open(name: PathType, mode: str = "r") -> Any:
+        """Open a file at the given path."""
+        raise NotImplementedError()
+
+    @staticmethod
+    def copyfile(src: PathType, dst: PathType, overwrite: bool = False) -> None:
```

Comment:
nitpick, but maybe say what setting overwrite to True/False does? Otherwise really nice docs